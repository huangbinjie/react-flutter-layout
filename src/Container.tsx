import React from 'react'
import { detect } from 'detect-browser'
import { Alignment, decodeAlignment } from './style'

const browser = detect()
const isIe = browser && browser.name === 'ie'

export interface IContainer {
  className?: string
  color?: string
  children?: React.ReactNode
  width?: string
  height?: string
  alignment?: Alignment
  margin?: string
  padding?: string
  shrinkWrap?: boolean
  style?: React.CSSProperties
}

/**
 * 容器组件，提供丰富的布局效果。
 * to honor the `width`, `height`,
 * to expand to fit the parent,
 * to be as big as possible.
 *
 * @param props
 */
export default function Container(props: IContainer) {
  const {
    color,
    width,
    height,
    alignment = 'start',
    margin,
    padding,
    shrinkWrap,
    style,
    children,
    ...restProps
  } = props

  return (
    <div
      {...restProps}
      style={{
        ...style,
        display: shrinkWrap ? isIe ? '-ms-inline-flex' : 'inline-flex' : 'flex',
        flex: shrinkWrap ? 'none' : 1,
        justifyContent: decodeAlignment(alignment),
        width,
        height,
        margin,
        padding,
        // flex 1 width 不生效，用约束让其生效
        maxWidth: width,
        maxHeight: height,
        backgroundColor: color
      }}
    >
      {children}
    </div>
  )
}
