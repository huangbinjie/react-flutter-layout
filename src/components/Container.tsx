import React from 'react'
import { Properties } from 'csstype'
import { detect } from 'detect-browser'
import { Alignment, decodeAlignment, LayoutStyle, LayoutStyleKey } from '../style'
import { Constraint } from '../constraint'

const browser = detect()
const isIe = browser && browser.name === 'ie'

export interface Container extends LayoutStyle {
  width?: string
  height?: string
  margin?: string
  padding?: string
  // 背景色
  color?: string
  alignment?: Alignment
  className?: string
  children?: React.ReactNode
  constraints?: Constraint
  debuggerLabel?: string
  style?: Omit<Properties, LayoutStyleKey | keyof Constraint>
}

/**
 * 容器组件，提供丰富的布局效果。
 * to honor the `width`, `height`,
 * to shrink to the children,
 * to be as small as possible.
 *
 * @param props
 */
export default function Container(props: Container) {
  const {
    color,
    width,
    height,
    alignment = 'start',
    margin,
    padding,
    style,
    children,
    debuggerLabel = 'Container',
    ...restProps
  } = props

  const shrinkWrap = children != void 0

  return (
    <div
      debugger-label={debuggerLabel}
      {...restProps}
      style={{
        ...style,
        display: shrinkWrap ? isIe ? '-ms-inline-flex' : 'inline-flex' : 'flex',
        flex: shrinkWrap ? 'none' : 1,
        justifyContent: decodeAlignment(alignment),
        // shrinkWrap height
        alignSelf: shrinkWrap ? 'baseline' : 'auto',
        width,
        height,
        margin,
        padding,
        // flex 1 width 不生效，用约束让其生效
        maxWidth: width,
        maxHeight: height,
        backgroundColor: color
      }}
    >
      {children}
    </div>
  )
}
