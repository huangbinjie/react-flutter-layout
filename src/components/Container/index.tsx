import React, { forwardRef, useContext } from 'react'
import { Properties } from 'csstype'
import { detect } from 'detect-browser'
import { style as jss } from 'typestyle'
import classnames from 'classnames'
import {
  Alignment,
  decodeAlignment,
  LayoutStyle,
  LayoutStyleKey,
  Overflow,
  DirectionContext
} from '../../style'
import { Constraint } from '../../constraint'
import { only, Omit } from '../../utils'

const browser = detect()
const isIe = browser && browser.name === 'ie'

export type ContainerGap = {
  top?: 0 | string
  left?: 0 | string
  bottom?: 0 | string
  right?: 0 | string
}

export interface ContainerProps extends Omit<LayoutStyle, 'margin' | 'padding'> {
  /**
   * 指定 Container 子节点的对其方式
   * 
   * 注意，Container 默认为 Block 布局，指定了该属性则会变成 Flex 布局
   */
  alignment?: Alignment
  width?: string
  height?: string
  margin?: string | ContainerGap
  padding?: string | ContainerGap
  /**
   * 背景色
   */
  color?: string
  className?: string
  children?: React.ReactNode
  /**
   * 约束
   * 
   * 对自己生效
   */
  constraints?: Constraint
  overflow?: Overflow
  style?: Omit<Properties, LayoutStyleKey | keyof Constraint>
}

/**
 * 容器组件，提供丰富的布局效果
 * 
 * to honor the `width`, `height`,
 * to shrink to the children,
 * to be as small as possible.
 * 
 * @param props
 */
export default forwardRef<HTMLDivElement, ContainerProps>(function Container(props, ref) {
  const {
    className,
    constraints = {},
    color,
    width,
    height,
    alignment,
    margin,
    padding,
    children,
    overflow = 'visible',
    ...restProps
  } = props

  const { maxWidth = 'none', maxHeight = 'none', minWidth = 'auto', minHeight = 'auto' } = constraints
  const direction = useContext(DirectionContext)
  const isAvailableWidth = width && width !== 'auto'
  const isAvailableHeight = height && height !== 'auto'

  // 有子组件会收缩，有高宽会收缩
  const shrinkWrap =
    children != void 0 ||
    (direction === 'row' && isAvailableWidth) ||
    (direction === 'column' && isAvailableHeight)

  let displayStyle: object

  if (alignment) {
    const flexStyle = {
      display: shrinkWrap ? (isIe ? '-ms-inline-flex' : 'inline-flex') : 'flex',
      flexGrow: shrinkWrap ? 0 : 1,
      flexShrink: isAvailableWidth || isAvailableHeight ? 0 : 1,
      justifyContent: decodeAlignment(alignment),
      // shrinkWrap height
      alignSelf: shrinkWrap ? 'baseline' : 'auto',
    }

    displayStyle = flexStyle
  } else {
    const blockStyle = {
      height: shrinkWrap ? 'auto' : '100%'
    }

    displayStyle = blockStyle
  }

  const classname = jss(displayStyle, {
    width,
    height,
    margin: decodeGap(margin),
    padding: decodeGap(padding),
    maxWidth: maxWidth,
    maxHeight: maxHeight,
    minWidth: minWidth,
    minHeight: minHeight,
    overflow,
    backgroundColor: color
  })

  return (
    <DirectionContext.Provider value="row">
      <div ref={ref} className={classnames(classname, className)} debug-label="Container" {...restProps}>
        {only(children)}
      </div>
    </DirectionContext.Provider>
  )
})

function decodeGap(margin?: string | ContainerGap): string | undefined {
  if (margin == null) return undefined
  if (typeof margin === 'string') {
    return margin
  }
  const { top = 0, right = 0, bottom = 0, left = 0 } = margin
  return `${top} ${right} ${bottom} ${left}`
}
