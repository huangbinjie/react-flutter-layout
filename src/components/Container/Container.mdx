---
name: Container
menu: components
---

import { Playground, Props } from 'docz'
import Container from '../Container'

## Container

容器。基础组件，提供各种布局效果。只能传一个子组件。

有子组件的时候，Container 会收缩紧贴子组件的高宽。没有子组件的时候，Container 会尽可能的大以占满父容器。

> 内部使用 flex 布局，搭配其他元素使用的时候需注意。

## 不同与 flutter 的点

个人觉得前端的实现需要结合前端的场景和实现难度和用户感受，不需要可以的照抄 flutter，所以总有一些不太一样的地方。

- 约束的生效方式受限于前端实现，只能对自己约束，而不是子组件。
- 为使 Container 能更胜任前端的日常工作，增加了类似 overflow 的属性。

## Usage

### Basic

<Playground>
  <Container width="100px" height="100px" color="#ececec">
    展示一个方块
  </Container>
</Playground>

### 有子组件

有子组件的时候， Container 会紧贴子组件。

<Playground>
  <Container color="#ececec">
    展示一个方块
  </Container>
</Playground>

### 没有子组件

没有子组件的时候，Container 会尽可能的大。

<Playground>
  <Container width="100px" height="100px">
    <Container color="#ececec"></Container>
  </Container>
</Playground>

### 约束

约束只对组件本身有效。

<Playground>
  <Container constraints={{maxWidth: '100px', maxHeight: '100px'}} width="200px" height="200px" color="#ececec">
    100 的方块
  </Container>
</Playground>

## Props

<Props of={Container} />
