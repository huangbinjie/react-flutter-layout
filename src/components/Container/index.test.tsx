import React from 'react'
import renderer from 'react-test-renderer'

import Container from '.'

describe('Container', () => {
  test('basic', () => {
    const tree = renderer.create(
      <Container>empty content</Container>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
