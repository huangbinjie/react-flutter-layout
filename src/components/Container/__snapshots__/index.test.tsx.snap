// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container basic 1`] = `
<div
  className="f1zge5p"
  debug-label="Container"
>
  empty content
</div>
`;
